Common error in SOLR logs

This is a common error in the solr logs,
I found about 4000 of these in one days worth of logs (old log: catalina.2018-02-13.log)

```
Feb 13, 2018 2:38:11 AM org.apache.solr.common.SolrException log
SEVERE: org.apache.solr.common.SolrException
        at org.apache.solr.handler.component.QueryComponent.prepare(QueryComponent.java:134)
        at org.apache.solr.handler.component.SearchHandler.handleRequestBody(SearchHandler.java:165)
        at org.apache.solr.handler.RequestHandlerBase.handleRequest(RequestHandlerBase.java:129)
        at org.apache.solr.core.SolrCore.execute(SolrCore.java:1376)
        at org.apache.solr.servlet.SolrDispatchFilter.execute(SolrDispatchFilter.java:365)
        at org.apache.solr.servlet.SolrDispatchFilter.doFilter(SolrDispatchFilter.java:260)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:235)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:233)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:191)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:127)
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:103)
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:109)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:293)
        at org.apache.coyote.http11.Http11Processor.process(Http11Processor.java:861)
        at org.apache.coyote.http11.Http11Protocol$Http11ConnectionHandler.process(Http11Protocol.java:606)
        at org.apache.tomcat.util.net.JIoEndpoint$Worker.run(JIoEndpoint.java:489)
        at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.lucene.queryParser.ParseException: Cannot parse '(author_name:() OR author_alternative_name:()) _val_:"sqrt(edition_count)"^10': Encountered " ")" ") "" at line 1, column 14.
Was expecting one of:
    <NOT> ...
    "+" ...
    "-" ...
    <BAREOPER> ...
    "(" ...
    "*" ...
    <QUOTED> ...
    <TERM> ...
    <PREFIXTERM> ...
    <WILDTERM> ...
    "[" ...
    "{" ...
    <NUMBER> ...
    
        at org.apache.lucene.queryParser.QueryParser.parse(QueryParser.java:211)
        at org.apache.solr.search.LuceneQParser.parse(LuceneQParserPlugin.java:79)
        at org.apache.solr.search.QParser.getQuery(QParser.java:143)
        at org.apache.solr.handler.component.QueryComponent.prepare(QueryComponent.java:105)
        ... 17 more
Caused by: org.apache.lucene.queryParser.ParseException: Encountered " ")" ") "" at line 1, column 14.
Was expecting one of:
    <NOT> ...
    "+" ...
    "-" ...
    <BAREOPER> ...
    "(" ...
    "*" ...
    <QUOTED> ...
    <TERM> ...
    <PREFIXTERM> ...
    <WILDTERM> ...
    "[" ...
    "{" ...
    <NUMBER> ...
    
        at org.apache.lucene.queryParser.QueryParser.generateParseException(QueryParser.java:1852)
        at org.apache.lucene.queryParser.QueryParser.jj_consume_token(QueryParser.java:1731)
        at org.apache.lucene.queryParser.QueryParser.Clause(QueryParser.java:1337)
        at org.apache.lucene.queryParser.QueryParser.Query(QueryParser.java:1245)
        at org.apache.lucene.queryParser.QueryParser.Clause(QueryParser.java:1323)
        at org.apache.lucene.queryParser.QueryParser.Query(QueryParser.java:1245)
        at org.apache.lucene.queryParser.QueryParser.Clause(QueryParser.java:1323)
        at org.apache.lucene.queryParser.QueryParser.Query(QueryParser.java:1245)
        at org.apache.lucene.queryParser.QueryParser.TopLevelQuery(QueryParser.java:1234)
        at org.apache.lucene.queryParser.QueryParser.parse(QueryParser.java:206)
        ... 20 moreif 

Feb 13, 2018 2:38:11 AM org.apache.solr.core.SolrCore execute
INFO: [] webapp=/solr path=/select params={spellcheck=true&facet=true&wt=json&rows=100&fl=*&start=0&q=(author_name:()+OR+author_alternative_name:())+_val_:"sqrt(edition_count)"^10&q.op=AND&spellcheck.count=3&facet.field=has_fulltext&facet.field=author_facet&facet.field=language&facet.field=first_publish_year&facet.field=publisher_facet&facet.field=subject_facet&facet.field=person_facet&facet.field=place_facet&facet.field=time_facet&facet.field=public_scan_b&fq=type:work} status=400 QTime=0 
```

Line that generates this: https://github.com/internetarchive/openlibrary/blob/5110112995c56593694c97f28b89e7eed659a920/openlibrary/plugins/worksearch/code.py#L155

when `v` is an  empty string this error occurs. Unfortunately I have not been able to reproduce this naturally with query strings

The code is triggered when the search path `/search?author=blah` is used, and somehow the author param ends up empty. In all my attempts via a browser / curl the author param is removed earlier if it evaluates to an empty string, so the failing code never executes. There must be some other way to end up there with `v = ''`

I was able to reproduce the error exactly by injecting an empty string into the code.

`build_q_list()` in only called by `run_solr_query()`, which is only called by `do_search()`, which is as far as I can see, only used by UI searches, so is not caused by the SOLR updater srcipts that I'm trying to fix.

--- a/openlibrary/plugins/worksearch/code.py
+++ b/openlibrary/plugins/worksearch/code.py
@@ -148,11 +148,14 @@ def build_q_list(param):
         if 'author' in param:
             v = param['author'].strip()
             m = re_author_key.search(v)
-            if m: # FIXME: 'OL123A OR OL234A'
-                q_list.append('author_key:(' + m.group(1) + ')')
+            if m:
+                q_list.append("author_key:(%s)" % m.group(1))
             else:
                 v = re_to_esc.sub(lambda m:'\\' + m.group(), v)
-                q_list.append('(author_name:(' + v + ') OR author_alternative_name:(' + v + '))')
+                # Somehow v can be empty at this point,
+                #   passing the following with empty strings causes a severe error in SOLR
+                if v:
+                    q_list.append("(author_name:(%(name)s) OR author_alternative_name:(%(name)s))" % {'name': v})
 
         check_params = ['title', 'publisher', 'oclc', 'lccn', 'contribtor', 'subject', 'place', 'person', 'time']
         q_list += ['%s:(%s)' % (k, param[k]) for k in check_params if k in param]
@@ -251,7 +254,7 @@ def run_solr_query(param = {}, rows=100, page=1, sort=None, spellcheck_count=Non
 
     solr_result = execute_solr_query(solr_select)
     if solr_result is None:
-        return (None, None, None)
+        return (None, solr_select, q_list)
     reply = solr_result.read()
     return (reply, solr_select, q_list)
 

